realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java

```diff
@@ -1947,6 +1949,95 @@ public void onChange(RealmResults<AllTypes> object) {
         result.load();
     }
 
+    // handlerController#emptyAsyncRealmObject is accessed from different threads
+    // make sure that we iterate over it safely without any race condition (ConcurrentModification)
+    @Test
+    @RunTestInLooperThread
+    public void concurrentModificationEmptyAsyncRealmObject() {
+        final Realm realm = looperThread.realm;
+        Dog dog1 = new Dog();
+        dog1.setName("Dog 1");
+
+        Dog dog2 = new Dog();
+        dog2.setName("Dog 2");
+
+        realm.beginTransaction();
+        dog1 = realm.copyToRealm(dog1);
+        dog2 = realm.copyToRealm(dog2);
+        realm.commitTransaction();
+
+        final WeakReference<RealmObjectProxy> weakReference1 = new WeakReference<RealmObjectProxy>((RealmObjectProxy)dog1);
+        final WeakReference<RealmObjectProxy> weakReference2 = new WeakReference<RealmObjectProxy>((RealmObjectProxy)dog2);
+
+        final RealmQuery<Dog> dummyQuery = RealmQuery.createQuery(realm, Dog.class);
+        looperThread.realm.handlerController.emptyAsyncRealmObject.put(weakReference1, dummyQuery);
```

Comment:
Since this is touching the internal implementation, perhaps add a comment  about when this is normally triggered, because emptyAsyncRealmObject is normally only used in conjunction with `findFirstAsync` right?